File: code/geometry/geometry.py
def vecsub(a, b):
	return (a[0] - b[0], a[1] - b[1])
def vecadd(a, b):
	return (a[0] + b[0], a[1] + b[1])
def dot(a, b):
	return a[0] * b[0] + a[1] * b[1]
def cross(a, b, o = (0, 0)):
	return (a[0] - o[0]) * (b[1] - o[1]) - (a[1] - o[1]) * (b[0] - o[0])
def len2(a):
	return a[0] ** 2 + a[1] ** 2
def dist2(a, b):
	return len2(vecsub(a, b))
def sign(x):
	return (x > 0) - (x < 0)
def zero(x):
	return abs(x) < 1E-9

File: code/geometry_cpp/geometry.cpp
template <typename T> struct point {
	T x, y;
	point() { x=y=0; }
	point(T xx, T yy) : x(xx), y(yy) { }
	template <typename U> point(point<U> o) : x(o.x), y(o.y) { }
	point operator+(point o) const { return { x+o.x, y+o.y }; }
	point operator-(point o) const { return { x-o.x, y-o.y }; }
	point operator*(T o) const { return { x*o, y*o }; }
	point operator/(T o) const { return { x/o, y/o }; }
	bool operator==(point o) const { return x==o.x && y==o.y; };
	bool operator<(point o) const { return tie(x, y) < tie(o.x, o.y); };
	T dot(point o) const { return x*o.x + y*o.y; }
	T cross(point b) const { return x*b.y - y*b.x; }
	T cross(point b, point o) const { return (*this-o).cross(b-o); }
	T len2() const { return x*x + y*y; }
};
using ipoint = point<ll>;
using dpoint = point<double>;
ll sign(auto x) { return (x>0) - (x<0); }
bool zero(double x) { return abs(x) < 1E-9; }
